File: packages/semantic-layer/src/adapter/ast-adapter.ts
/**
 * ASTAdapter — SLA 코어와 AST JSON 사이의 인터페이스 경계.
 *
 * SLA 엔진은 이 인터페이스를 통해서만 AST 데이터에 접근한다.
 * AST JSON 구조가 변경되면 구현체(ASTJsonAdapter)만 수정.
 */

import type {
  PageInfo,
  BlockInfo,
  ParagraphInfo,
  StoryInfo,
  StyleInfo,
  ImageInfo,
  TableInfo,
  FontInfo,
  ColorMap,
} from './types.js'

export interface ASTAdapter {
  // 문서 구조
  getPages(): PageInfo[]
  getBlocks(pageNumber: number): BlockInfo[]
  getParagraphs(blockId: string): ParagraphInfo[]

  // 스토리
  getStories(): StoryInfo[]
  getStory(storyId: string): StoryInfo | null
  getFrameIdsForStory(storyId: string): string[]

  // 스타일
  getParagraphStyles(): StyleInfo[]
  getCharacterStyles(): StyleInfo[]
  getStyleByRef(ref: string): StyleInfo | null

  // 미디어
  getImageInfo(blockId: string): ImageInfo | null
  getTableInfo(blockId: string): TableInfo | null

  // 폰트
  getFonts(): FontInfo[]

  // 색상
  getColors(): ColorMap
  getColorHex(colorRef: string): string | null

  // 메타
  getSourceFile(): string | null
  getDocumentHash(): string
}

File: src/main/java/kr/dogfoot/hwpxlib/tool/idmlconverter/semantic/adapter/ASTAdapter.java
package kr.dogfoot.hwpxlib.tool.idmlconverter.semantic.adapter;

import java.util.List;

/**
 * ASTAdapter — SLA 코어와 AST 사이의 인터페이스 경계.
 *
 * <p>FeatureExtractor / RuleEngine / RelationBuilder 는 이 인터페이스를
 * 통해서만 AST 데이터에 접근한다. 구체 구현은 ASTDocument 객체 또는
 * AST JSON 둘 다 가능.</p>
 *
 * <p>TS {@code ASTAdapter} 인터페이스와 1:1.</p>
 */
public interface ASTAdapter {
    // 문서 구조
    List<PageInfo> getPages();
    List<BlockInfo> getBlocks(int pageNumber);
    List<ParagraphInfo> getParagraphs(String blockId);

    // 스토리
    List<StoryInfo> getStories();
    StoryInfo getStory(String storyId);

    // 스타일
    StyleInfo getStyleByRef(String ref);

    // 메타
    String getDocumentHash();
}
